e2e/e2e_test.go

```diff
@@ -0,0 +1,301 @@
+		}
+	}
+	if len(missingEnvs) > 0 {
+		t.Fatalf("missing required environment variable(s): %s", missingEnvs)
+	}
+
+	s3Cl, err := createS3Client()
+	if err != nil {
+		t.Fatalf("failed to create S3 client: %s", err)
+	}
+
+	tests := []struct {
+		desc    string
+		kubeVer string
+	}{
+		{
+			desc:    "latest release",
+			kubeVer: "1.12.0",
+		},
+		{
+			desc:    "previous release",
+			kubeVer: "1.11.2",
+		},
+		{
+			desc:    "previous previous release",
+			kubeVer: "1.10.6",
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.desc, func(t *testing.T) {
+			l := log.New(os.Stdout, fmt.Sprintf("[%s] ", t.Name()), 0)
+			dnsName := toDNSName(t.Name())
+
+			wd, err := os.Getwd()
+			if err != nil {
+				t.Fatalf("failed to get working directory: %s", err)
+			}
+			kubeConfFile := path.Join(wd, "kubeconfig-e2e."+dnsName)
+
+			// Delete old kubeconfig
+			if err := os.Remove(kubeConfFile); err != nil && !os.IsNotExist(err) {
+				t.Fatalf("failed to delete kubeconfig %q: %s", kubeConfFile, err)
+			}
+
+			// Create space.
+			storeName := toS3Name(fmt.Sprintf("%s-%s", os.Getenv(kopsEnvVarClusterName), dnsName))
+			if err := s3Cl.deleteSpace(storeName); err != nil {
+				t.Fatalf("failed to delete space %q (pre-test): %s", storeName, err)
+			}
+			if err := s3Cl.ensureSpace(storeName); err != nil {
+				t.Fatalf("failed to ensure space %q: %s", storeName, err)
+			}
+			defer func() {
+				if err := s3Cl.deleteSpace(storeName); err != nil {
+					t.Fatalf("failed to delete space %q (post-test): %s", storeName, err)
+				}
+			}()
+
+			// Create cluster.
+			extraEnvs := []string{
+				fmt.Sprintf("%s=do://%s", kopsEnvVarStateStore, storeName),
+				"KUBECONFIG=" + kubeConfFile,
+			}
+			if err := runScript(extraEnvs, "destroy_cluster.sh"); err != nil {
+				t.Fatalf("failed to destroy cluster (pre-test): %s", err)
+			}
+			if err := runScript(extraEnvs, "setup_cluster.sh", tt.kubeVer, strconv.Itoa(numWantNodes)); err != nil {
+				t.Fatalf("failed to set up cluster: %s", err)
+			}
+			defer func() {
+				if err := runScript(extraEnvs, "destroy_cluster.sh"); err != nil {
+					t.Errorf("failed to destroy cluster (post-test): %s", err)
+				}
+			}()
+
+			cs, err := kubeClient(kubeConfFile)
+			if err != nil {
+				t.Fatalf("failed to create Kubernetes client: %s", err)
+			}
+
+			// Check that nodes become ready.
+			l.Println("Polling for node readiness")
+			var (
+				gotNodes      []corev1.Node
+				numReadyNodes int
+			)
+			start := time.Now()
+			if err := wait.Poll(5*time.Second, 6*time.Minute, func() (bool, error) {
+				nl, err := cs.Core().Nodes().List(metav1.ListOptions{LabelSelector: "kubernetes.io/role=node"})
+				if err != nil {
+					return false, err
+				}
+
+				gotNodes = nl.Items
+				numReadyNodes = 0
+				for _, node := range gotNodes {
+					for _, cond := range node.Status.Conditions {
+						if cond.Type == corev1.NodeReady && cond.Status == corev1.ConditionTrue {
```

Comment:
I think it's really important here that we also verify that there are no taints on the node to test registration, particularly the cloud taint `node.cloudprovider.kubernetes.io/uninitialized` should not exist.

Reply:
I pushed a commit that makes sure the taint you mentioned isn't set. Let me know if you want me to  broaden it further to "no taints at all must be set".